new-components/reporters/json-logger/cmd/main_test.go

```diff
@@ -0,0 +1,317 @@
+package main
+
+import (
+	"context"
+	"log/slog"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/types/known/structpb"
+
+	"github.com/smithy-security/smithy/sdk/component"
+	ocsf "github.com/smithy-security/smithy/sdk/gen/com/github/ocsf/ocsf_schema/v1"
+)
+
+type (
+	capturingLogger struct {
+		logs   []string
+		fields []slog.Attr
+	}
+
+	store struct {
+		vulns []*ocsf.VulnerabilityFinding
+	}
+)
+
+func (s *store) Close(ctx context.Context) error {
+	return nil
+}
+
+func (s *store) Validate(finding *ocsf.VulnerabilityFinding) error {
+	return nil
+}
+
+func (s *store) Read(ctx context.Context, instanceID string) ([]*ocsf.VulnerabilityFinding, error) {
+	return s.vulns, nil
+}
+
+func (s *store) Update(ctx context.Context, instanceID string, findings []*ocsf.VulnerabilityFinding) error {
+	return nil
+}
+
+func (s *store) Write(ctx context.Context, instanceID string, findings []*ocsf.VulnerabilityFinding) error {
+	return nil
+}
+
+func (c *capturingLogger) capture(msg string, keyvals ...any) {
+	if msg != "" {
+		c.logs = append(c.logs, msg)
+	}
+
+	if len(keyvals) > 0 {
+		newAttr := slog.Attr{}
+		for idx, kv := range keyvals {
+			attr, ok := kv.(slog.Attr)
+			if ok {
+				c.fields = append(c.fields, attr)
+				continue
+			}
+
+			if idx%2 == 0 {
+				newAttr.Key = kv.(string)
+				continue
+			}
+
+			newAttr.Value = slog.AnyValue(kv)
+			c.fields = append(c.fields, newAttr)
+		}
+	}
+}
+
+func (c *capturingLogger) Debug(msg string, keyvals ...any) {
+	c.capture(msg, keyvals...)
+}
+
+func (c *capturingLogger) Info(msg string, keyvals ...any) {
+	c.capture(msg, keyvals...)
+}
+
+func (c *capturingLogger) Warn(msg string, keyvals ...any) {
+	c.capture(msg, keyvals...)
+}
+
+func (c *capturingLogger) Error(msg string, keyvals ...any) {
+	c.capture(msg, keyvals...)
+}
+
+func (c *capturingLogger) With(args ...any) component.Logger {
+	c.capture("", args...)
+	return c
+}
+
+func ptr[T any](v T) *T {
```

Comment:
you can use ptr.Ptr instead of this

Reply:
As mentioned in another comment, here we only need this in the test